### PR TITLE
Model agnosticism

### PR DESCRIPTION
**Pull request by jxmorris12**
- change ":" in command-line API to "|" (should hopefully make paths work on Windows)
- add back `distilbert-base-uncased-mr` model
- remove commented-out imports all over the place
- remove `try_model_predict` and other messy PyTorch model prediction guessing
- create `ModelWrapper` abstract class that models can implement to be used for attacks (this will allow us to add models from TensorFlow, sklearn, etc. with ease)

### CODE REVIEW

File: textattack/models/wrappers/pytorch_model_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import textattack
+
+from .model_wrapper import ModelWrapper
+
+
+class PyTorchModelWrapper(ModelWrapper):
+    """Loads a PyTorch model (`nn.Module`) and tokenizer."""
+
+    # TODO test training code before putting up PR
```
Comment (uvafan) [inline textattack/models/wrappers/pytorch_model_wrapper.py:11]: Has this TODO been done?

File: textattack/models/wrappers/pytorch_model_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                f"PyTorch model must be torch.nn.Module, got type {type(model)}"
+            )
+
+        self.model = model.to(textattack.shared.utils.device)
+        self.tokenizer = tokenizer
+        self.batch_size = batch_size
+
+    @property
+    def get_model_device(self):
+        if hasattr(self.model, "model"):
```
Comment (uvafan) [inline textattack/models/wrappers/pytorch_model_wrapper.py:24]: can you add a comment explaining why you have to do this? Why do some models have model attributes?

File: textattack/shared/attack.py [hunk 3]
```diff
@@ -166,6 +164,10 @@ def filter_transformations(
             current_text: The current ``AttackedText`` on which the transformation was applied.
             original_text: The original ``AttackedText`` from which the attack started.
         """
+        # Remove any occurences of current_text in transformed_texts
```
Comment (uvafan) [inline textattack/shared/attack.py:167]: In which cases do we need to do this? I feel like this should be up to the transformation not to do?

File: textattack/shared/attacked_text.py [hunk 4]
```diff
@@ -346,7 +346,11 @@ def generate_new_attacked_text(self, new_words):
     @property
     def tokenizer_input(self):
         """The tuple of inputs to be passed to the tokenizer."""
-        return tuple(self._text_input.values())
+        input_tuple = tuple(self._text_input.values())
+        # Prefer to return a string instead of a tuple with a single value.
+        if len(input_tuple) == 1:
+            input_tuple = input_tuple[0]
```
Comment (uvafan) [inline textattack/shared/attacked_text.py:352]: nit: I would change this to `return input_tuple[0]`.

File: textattack/shared/utils/misc.py [hunk 5]
```diff
@@ -57,34 +56,41 @@ def html_table_from_rows(rows, title=None, header=None, style_dict=None):
...
+    model_cache_path = textattack.shared.utils.download_if_needed(model_path)
+    train_args_path = os.path.join(model_cache_path, "train_args.json")
+    if not os.path.exists(train_args_path):
+        textattack.shared.logger.warn(
+            f"train_args.json not found in model path {model_path}. Defaulting to 2 labels."
+        )
+        return 2
+    else:
+        args = json.loads(open(train_args_path).read())
+        return args["num_labels"]
```
Comment (uvafan) [inline textattack/shared/utils/misc.py:71]: Should we include a default to 2 similar to above here if `num_labels` isn't a key?

File: textattack/shared/utils/tensor.py [hunk 6]
```diff
@@ -1,119 +1,31 @@
+import numpy as np
 import torch
 
-import textattack
 
-# from textattack.shared import utils
+def batch_model_predict(model_predict, inputs, batch_size=32):
```
Comment (uvafan) [inline textattack/shared/utils/tensor.py:5]: Wow this seems much nicer now. I assume it still works same as before?

File: textattack/transformations/word_swap_gradient_based.py [hunk 7]
```diff
@@ -18,8 +19,16 @@ class WordSwapGradientBased(Transformation):
...
     """
 
-    def __init__(self, model, top_n=1):
+    def __init__(self, model_wrapper, top_n=1):
+        # Unwrap model wrappers. Need raw model for gradient.
+        if not isinstance(model_wrapper, textattack.models.wrappers.ModelWrapper):
+            raise TypeError(f"Got invalid model wrapper type {type(model_wrapper)}")
+        model = model_wrapper.model
+        self.model_wrapper = model_wrapper
+        self.model = model
```
Comment (uvafan) [inline textattack/transformations/word_swap_gradient_based.py:28]: nit: combine this and line 26 into `self.model = model_wrapper.model`

Comment (uvafan) [reply textattack/transformations/word_swap_gradient_based.py:28]: Also maybe set a `self.tokenizer` so you only have to change it in one place if you change it again?

File: textattack/transformations/word_swap_inflections.py [hunk 8]
```diff
@@ -35,7 +35,7 @@ def _get_replacement_words(self, word, word_part_of_speech):
 
     def _get_transformations(self, current_text, indices_to_modify):
         words = current_text.words
-        sentence = Sentence(current_text.text)
+        sentence = Sentence(" ".join(words))
```
Comment (uvafan) [inline textattack/transformations/word_swap_inflections.py:38]: Why this change?

File: textattack/models/wrappers/pytorch_model_wrapper.py (hunk 1 above)
Comment (jxmorris12) [reply textattack/models/wrappers/pytorch_model_wrapper.py:11]: yes :)

File: textattack/shared/utils/tensor.py (hunk 6 above)
Comment (jxmorris12) [reply textattack/shared/utils/tensor.py:5]: yep (all the tests pass anyway)

File: textattack/models/wrappers/pytorch_model_wrapper.py (hunk 2 above)
Comment (jxmorris12) [reply textattack/models/wrappers/pytorch_model_wrapper.py:24]: it's just a handy thing to have for pytorch models. I added it bc it was a helper that was already in the code. Since we don't use it, though, let's get rid of it.

File: textattack/shared/attack.py (hunk 3 above)
Comment (jxmorris12) [reply textattack/shared/attack.py:167]: I think this is a good idea. It's a bug that's appeared maybe 4 times in 4 separate transformations (most recently, BERT masked-LM swap) – and identity transformations break the search methods

File: textattack/shared/utils/misc.py (hunk 5 above)
Comment (jxmorris12) [reply textattack/shared/utils/misc.py:71]: yes, good idea

File: textattack/transformations/word_swap_inflections.py (hunk 8 above)
Comment (jxmorris12) [reply textattack/transformations/word_swap_inflections.py:38]: FLAIR applies part-of-speech tags to punctuation and bits of punctuation. Without this, the part-of-speech tags wouldn't line up with `current_text.words` (and an error would be thrown in special cases)

File: textattack/transformations/word_swap_gradient_based.py (hunk 7 above)
Comment (jxmorris12) [reply textattack/transformations/word_swap_gradient_based.py:28]: good idea

File: textattack/commands/attack/attack_args_helpers.py [hunk 9]
```diff
@@ -18,6 +18,10 @@
     WHITE_BOX_TRANSFORMATION_CLASS_NAMES,
 )
 
+# The split token allows users to optionally pass multiple arguments in a single
+# parameter by separating them with the split token.
+ARGS_SPLIT_TOKEN = "|"
```
[the remark this answers is not in the document]
Comment (jxmorris12) [reply textattack/commands/attack/attack_args_helpers.py:23]: yeah, don't think we can do much more unfortunately

File: textattack/shared/attack.py (hunk 3 above)
Comment (uvafan) [reply textattack/shared/attack.py:167]: Can we throw an error if there's a duplicate then? Rather than quietly filtering it out?

File: textattack/transformations/word_swap_inflections.py (hunk 8 above)
Comment (uvafan) [reply textattack/transformations/word_swap_inflections.py:38]: Hmm can it assign the right POS to words without punctuation though?

Comment (jxmorris12) [reply textattack/transformations/word_swap_inflections.py:38]: yeah, this is a good concern, I shall put it on Asana

File: textattack/shared/attack.py (hunk 3 above)
Comment (jxmorris12) [reply textattack/shared/attack.py:167]: done
